enhancements/20220823-node-manager-support-ksm.md

```diff
@@ -0,0 +1,159 @@
+
+It writes to the KSM file as configured, gets the node's memory usage at regular intervals, and starts or stops the ksmd program according to the threshold calculation.
+
+**Ksmtuned CR**
+
+Configuration of management node ksmtuned via Ksmtuned CR.
+
+Use `node.harvesterhci.io/ksmtuned-mode` in Annotations to save the selected modes.
+
+```yaml
+apiVersion: harvesterhci.io/v1beta1
+kind: Ksmtuned
+metadata:
+  annotations: 
+    node.harvesterhci.io/ksmtuned-mode: standard
+  name: harvester-master
+spec:
+  boost: 300
+  decay: -50
+  maxPages: 2000
+  minPages: 100
+  sleepMsec: 20
+  thresCoef: 20
+```
+
+Ksmtuned controller checks that the node's Ksmtuned CR does not exist, then it is automatically created for that node.
+
+Ksmtuned CR’s status to display [KSM metrics](https://www.kernel.org/doc/html/latest/admin-guide/mm/ksm.html#ksm-daemon-sysfs-interface) and ksmd phase.
+
+```yaml
+apiVersion: harvesterhci.io/v1beta1
+kind: Ksmtuned
+metadata:
+  annotations:
+    node.harvesterhci.io/ksmtuned-mode: customized
+  name: harvester-master
+spec:
+  ...
+status:
+  sharing: 17299179
+  shared: 312514
+  unshared: 2334736
+  volatile: 65571 
+  fullScans: 14102
+  stableNodeChains: 30
+  stableNodeDups: 56249
+  ksmdPhase: Running/Stop
+```
+
+**Ksmtuned script**
+
+Since KSM runs in a single mode, it is possible for large memory devices to occupy the CPU for a long time. To solve this problem, you can set the ksmtuned script threshold to run KSM only when the free memory is less than a certain percentage.
+
+Ksmtuned CR parameters：
+
+- **thresCoef**: Free memory threshold, KSM starts when the memory available is less than the set percentage ratio, otherwise KSM stops.
+- **boost**: Increment the number of pages scanned at a time, incrementing if the memory availability is less than the threshold.
+- **decay**: Decrement the number of pages scanned at a time, decrementing if memory availability is greater than the threshold.
+- **maxPages**: Maximum number of pages per scan.
+- **minPages**: The minimum number of pages per scan, i.e. the configuration for the first run.
+- **sleepMsec**: Each time you change the configuration interval, calculate the interval between two scans by the formula.(`sleepMsec * 16 * 1024 * 1024 / total memory` ).
+
+**Node Controller**
+
+- When synchronizing a Node, if the Ksmtuned CR does not exist, it is automatically created for that node.
+- When the Node is deleted, the Ksmtuned CR of the node is deleted.
+
+**Ksmtuned Controller**
+
+- When Ksmtuned CR is created, the first configuration information and write file is recorded and the ksmtuned program is started.
+- When synchronizing Ksmtuned CR again, if the configuration has not changed from the last recorded configuration, return directly, otherwise re-write the configuration file, and re-run the ksmtuned program.
+- Terminates the ksmtuned program when Ksmtuned CR is deleted.
+
```

Comment:
A mutate webhook  for Ksmtuned  resources is required to automatically adjust the ksmtuned parameters based on the mode selected by the user

Reply:
or we can use :
```yaml
apiVersion: harvesterhci.io/v1beta1
kind: Ksmtuned
metadata:
  name: harvester-master
spec:
  mode： standard
```

```yaml
apiVersion: harvesterhci.io/v1beta1
kind: Ksmtuned
metadata:
  name: harvester-master
spec:
  mode： customized
  parameters:
    boost: 300
    decay: -50
    maxPages: 2000
    minPages: 100
    sleepMsec: 20
    thresCoef: 20
```
instead